corehq/form_processor/backends/couch.py

```diff
@@ -0,0 +1,78 @@
+from couchdbkit import ResourceNotFound
+
+from couchforms.models import doc_types, XFormInstance, XFormError
+from couchforms.exceptions import UnexpectedDeletedXForm
+from dimagi.utils.couch.undo import DELETED_SUFFIX
+
+from ..exceptions import XFormNotFound
+
+
+class XFormCouch(object):
+
+    @staticmethod
+    def create_from_generic(generic_xform, generic_attachment=None):
+        xform = XFormInstance.from_generic(generic_xform)
+        xform.save()
+        if generic_attachment:
+            xform.put_attachment(**generic_attachment.to_json())
+        return xform
+
+    @staticmethod
+    def get_attachment(xform_id, attachment_name):
+        return XFormInstance.get_db().fetch_attachment(xform_id, attachment_name)
+
+    @classmethod
+    def get_attachments(cls, xform_id):
+        xform = cls._get_xform(xform_id)
+        return xform.attachments
+
+    @classmethod
+    def archive(cls, xform_generic, user=None):
+        xform = cls._get_xform(xform_generic.id)
+        return xform.archive(user=user)
+
+    @classmethod
+    def unarchive(cls, xform_generic, user=None):
+        xform = cls._get_xform(xform_generic.id)
+        return xform.unarchive(user=user)
+
+    @classmethod
+    def get_xml_element(cls, xform_generic):
+        xform = cls._get_xform(xform_generic.id)
```

Comment:
The SQL models will need an attachments property to keep track of
attachments (not just the form but images etc). It should be an additional
JSON column with a similar data to the couch _attachments dict (name,
content_type, length, maybe hash)

In this specific case I don't think you need to get the form for the
_attachments property since you already know the name of the attachment
('form.xml')

On 23 October 2015 at 19:24, Ben <notifications@github.com> wrote:

> In corehq/form_processor/backends/couch.py
> <https://github.com/dimagi/commcare-hq/pull/8780#discussion_r42890791>:
>
> > +        xform = cls._get_xform(xform_id)
> > +        return xform.attachments
> > +
> > +    @classmethod
> > +    def archive(cls, xform_generic, user=None):
> > +        xform = cls._get_xform(xform_generic.id)
> > +        return xform.archive(user=user)
> > +
> > +    @classmethod
> > +    def unarchive(cls, xform_generic, user=None):
> > +        xform = cls._get_xform(xform_generic.id)
> > +        return xform.unarchive(user=user)
> > +
> > +    @classmethod
> > +    def get_xml_element(cls, xform_generic):
> > +        xform = cls._get_xform(xform_generic.id)
>
> this one also depends on the _attachments property which i had been
> assuming won't exist for the sql model
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/8780/files#r42890791>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa